### PR TITLE
feat: constrain custom conversation handler event version

### PR DESCRIPTION
**Pull request by sobolk**
#### Description of changes

Change how `@conversation` directive accepts custom conversation handler.

From:
```
@conversation(
  aiModel: "anthropic.claude-3-haiku-20240307-v1:0", 
  systemPrompt: "Hello, world!",
  functionName: "FnCustomHandlerChatBot",
  eventVersion: '1.0'
)
```

To:
```
@conversation(
  aiModel: "anthropic.claude-3-haiku-20240307-v1:0", 
  systemPrompt: "Hello, world!",
  handler: {
    functionName: "FnCustomHandlerChatBot",
    eventVersion: '1.0'
  }
)
```

While still accepting older version. Older version will be removed upon GA.

#### Motivation

This is a companion PR to https://github.com/aws-amplify/amplify-backend/pull/2089 .
And a pre-requisite for https://github.com/aws-amplify/amplify-api-next/pull/369 .

These PRs setup a versioning for the event protocol between conversation route resolvers and lambda handler.

The protocol versioning works as follows:
1. We plumb event version through custom conversation handler -> schema -> transformer at runtime.
2. We're using subset of semver, i.e. `<major>.<minor>`. Patch is meaningless for protocol versioning.
3. We assert that protocol version must match on `<major>` digit. (mismatches within minor version should result in disabled functionality, not runtime errors).

This prepares us for future handling of `2.x` and higher versions of event. Where we can:
1. Either create a payload compatible to provided custom handler.
2. Or fail with validation and ask customer to provide matching version of custom handler.

(This decision will be made later, when break actually happens, in this PR we establish necessary primitives to handle it in the future, and assert correctness of current state where we only have single major version).

##### CDK / CloudFormation Parameters Changed

NONE.

#### Description of how you validated changes

Added unit tests.

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] E2E test run linked
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] Any CDK or CloudFormation parameter changes are called out explicitly

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codebuild_specs/scripts/lint_pr.sh [hunk 1]
```diff
@@ -9,6 +9,8 @@ if [ -z "$PR_NUM" ]; then
   exit
 fi
 
+export NODE_OPTIONS=--max-old-space-size=4096
```
Comment (sobolk) [inline codebuild_specs/scripts/lint_pr.sh:12]: image

File: packages/amplify-graphql-conversation-transformer/src/__tests__/amplify-graphql-conversation-transformer.test.ts [hunk 2]
```diff
@@ -12,6 +12,7 @@ import * as path from 'path';
 import { Code, Function, IFunction, Runtime } from 'aws-cdk-lib/aws-lambda';
 import { GenerationTransformer } from '@aws-amplify/graphql-generation-transformer';
 import { toUpper } from 'graphql-transformer-common';
+import { describe } from 'jest-circus';
```
Comment (p5quared) [inline packages/amplify-graphql-conversation-transformer/src/__tests__/amplify-graphql-conversation-transformer.test.ts:15]: Is this necessary?

**Review by p5quared — APPROVED** [review]

File: packages/amplify-graphql-conversation-transformer/src/__tests__/amplify-graphql-conversation-transformer.test.ts (hunk 2 above)
Comment (sobolk) [reply packages/amplify-graphql-conversation-transformer/src/__tests__/amplify-graphql-conversation-transformer.test.ts]: not intentional, removed, thanks

**Review by p5quared — DISMISSED** [review]

**Review by atierian — DISMISSED** [review]
LGTM

I requested a review from @palpatim since this changes the public API of the `@conversation` directive.

File: packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts [hunk 3]
```diff
@@ -109,7 +110,18 @@ export class ConversationResolverGenerator {
...
+    if (directive.handler && directive.functionName) {
+      throw new Error("'functionName' and 'handler' are mutually exclusive");
+    }
+    if (directive.handler) {
+      const eventVersion = semver.coerce(directive.handler.eventVersion);
+      if (eventVersion?.major !== 1) {
+        throw new Error(
+          `Unsupported custom conversation handler. Expected eventVersion to match 1.x, received ${directive.handler.eventVersion}`,
+        );
+      }
```
Comment (palpatim) [inline packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts:122]: This seems more appropriate to house in the transformer's `validate` method.

Comment (sobolk) [reply packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts:122]: I had a feeling this is too deep in the implementation details, but couldn't find better place.
I'll take a stab at `validate` method, thanks for the pointer.

Comment (sobolk) [reply packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts]: done.

**Review by palpatim — APPROVED** [review]

**Review by atierian — APPROVED** [review]

**Review by atierian — APPROVED** [review]

**Review by p5quared — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]

**Review by palpatim — APPROVED** [review]

**Pull request merged**